tests/extension/oneapi_memcpy2d/memcpy2d_common.h

```diff
@@ -113,50 +121,55 @@ bool check_device_aspect_allocations(sycl::queue& queue) {
   }
 }
 
+/**
+ * @brief Function to check if the device can allocate memory according to the
+ * SrcPtrType and DestPtrType pointer types
+ */
 template <pointer_type SrcPtrType, pointer_type DestPtrType>
 bool check_device_aspect_allocations(sycl::queue& queue) {
   return check_device_aspect_allocations<SrcPtrType>(queue) &&
          check_device_aspect_allocations<DestPtrType>(queue);
 }
 
 /**
- * @brief Function for allocate memory
+ * @brief Function to allocate memory with type according to PtrType
  */
 template <typename T, pointer_type PtrType>
 auto allocate_memory(size_t size, sycl::queue& queue) {
   if constexpr (pointer_type::host == PtrType) {
-    auto deleter = [&](T* ptr) { delete[] ptr; };
-    return std::unique_ptr<T[], decltype(deleter)>(new T[size], deleter);
-  } else if constexpr (pointer_type::usm_host == PtrType) {
-    auto deleter = [&](T* ptr) { sycl::free(ptr, queue); };
-    return std::unique_ptr<T[], decltype(deleter)>(
-        sycl::malloc_host<T>(size, queue), deleter);
-  } else if constexpr (pointer_type::usm_device == PtrType) {
-    auto deleter = [&](T* ptr) { sycl::free(ptr, queue); };
-    return std::unique_ptr<T[], decltype(deleter)>(
-        sycl::malloc_device<T>(size, queue), deleter);
+    return std::make_unique<T[]>(size);
   } else {
-    auto deleter = [&](T* ptr) { sycl::free(ptr, queue); };
-    return std::unique_ptr<T[], decltype(deleter)>(
-        sycl::malloc_shared<T>(size, queue), deleter);
+    auto deleter = [=](T* ptr) { sycl::free(ptr, queue); };
+    if constexpr (pointer_type::usm_host == PtrType) {
+      return std::unique_ptr<T[], decltype(deleter)>(
+          sycl::malloc_host<T>(size, queue), deleter);
```

Comment:
I do not think you care about returning a `T[]` under the hood. So just trust the CTAD in all the 3 following unique pointers:
```suggestion
      return std::unique_ptr { sycl::malloc_host<T>(size, queue), deleter };
```

Reply:
I 'd like to do that, but the c++ specification doesn't allow that for constructors with custom deleters:
[unique_ptr](https://en.cppreference.com/w/cpp/memory/unique_ptr/unique_ptr)
> template< class U > unique_ptr( U p, /* see below */ d1 ) noexcept; | (3) | (constexpr since C++23)
> template< class U > unique_ptr( U p, /* see below */ d2 ) noexcept; | (4) | (constexpr since C++23)

> 3-4) Constructs a std::unique_ptr object which owns p, initializing the stored pointer with p and initializing a deleter D
> ....
> These two constructors are not selected by [class template argument deduction](https://en.cppreference.com/w/cpp/language/class_template_argument_deduction).